File: ros2/actuator_subscriber.cc
#include <list>
#include <memory>
#include <set>
#include <stdexcept>
#include <thread>
#include <vector>

#include "config/proto/config.pb.h"
#include "rclcpp/rclcpp.hpp"
#include "robot/action/factory/action_factory.h"
#include "robot/action/proto/action_packet.pb.h"
#include "ros2/node_runner.h"
#include "ros2/proto/ros2_data_type.pb.h"
#include "ros2/utils/packet_parser.h"

class ActionSubscriber : public rclcpp::Node {
 private:
  struct Actuator {
    std::string topic;
    std::shared_ptr<robot::action::ActionInterface> interface;
    std::pair<float, float> limits;
    rclcpp::SubscriptionBase::SharedPtr subscription;
    robot::action::ActionPacket reusable_packet;
  };

 public:
  ActionSubscriber(const std::string& node_name, const int node_id, const config::Config& config)
      : Node(node_name) {
    for (const auto& single_action : config.robot().actions().single_actions()) {
      if (single_action.action_type() != robot::action::ActionType::ACTUATOR ||
          static_cast<int>(single_action.node().id()) != node_id) {
        continue;
      }

      const auto& action_proto = single_action.actuator();

      auto interface =
          robot::action::ActionFactory::CreateAction(single_action, config.robot().boards());
      if (!interface.ok()) throw std::runtime_error(interface.status().ToString());

      auto shared_interface =
          std::shared_ptr<robot::action::ActionInterface>(std::move(interface.value()));

      robot::action::ActionPacket enable_packet;
      enable_packet.set_preset(robot::action::PresetCommand::PRESET_ENABLE_TORQUE);
      const auto enabled = shared_interface->SetAction(enable_packet);
      if (!enabled.ok()) throw std::runtime_error(enabled.ToString());

      for (const auto& subscription : single_action.node().subscriptions()) {
        const std::string& topic = subscription.topic();
        Actuator& actuator =
            actuators_.emplace_back(Actuator{.topic = topic,
                                             .interface = shared_interface,
                                             .limits = {action_proto.operational_lower_limit(),
                                                        action_proto.operational_upper_limit()}});

        auto result = ros2_utils::CreateActionMessageSubscription(
            *this,
            subscription,
            single_action,
            [this, &actuator](absl::StatusOr<robot::action::ActionPacket> parsed) {
              if (!parsed.ok()) {
                RCLCPP_ERROR(get_logger(),
                             "Invalid command on '%s': %s",
                             actuator.topic.c_str(),
                             parsed.status().ToString().c_str());
                return;
              }
              actuator.reusable_packet = *parsed;
              const auto [lower, upper] = actuator.limits;
              ros2_utils::DenormalizeActionPacket(actuator.reusable_packet, lower, upper);
              const auto status = actuator.interface->SetAction(actuator.reusable_packet);
              if (!status.ok()) {
                RCLCPP_ERROR(get_logger(),
                             "Actuator '%s' rejected command: %s",
                             actuator.topic.c_str(),
                             status.ToString().c_str());
              }
            });
        if (!result.ok()) throw std::invalid_argument(result.status().ToString());
        actuator.subscription = *result;
      }
    }

    if (actuators_.empty()) {
      RCLCPP_ERROR(
          this->get_logger(), "No actuators found in configuration for node_id %d!", node_id);
      return;
    }

    RCLCPP_INFO(this->get_logger(),
                "Actuator subscriber node started with %zu subscriptions for node_id %d!",
                actuators_.size(),
                node_id);
  }

  ~ActionSubscriber() {
    std::vector<std::thread> threads;

    std::set<robot::action::ActionInterface*> torn_down;
    for (auto& actuator : actuators_) {
      if (!torn_down.insert(actuator.interface.get()).second) continue;
      threads.emplace_back([&actuator]() {
        robot::action::ActionPacket teardown_packet;
        teardown_packet.set_preset(robot::action::PresetCommand::PRESET_TEARDOWN);
        auto status = actuator.interface->SetAction(teardown_packet);
        if (!status.ok()) {
          RCLCPP_ERROR(get_logger(), "Failed to teardown actuator '%s'", actuator.topic.c_str());
        }
      });
    }

    for (auto& thread : threads) {
      thread.join();
    }
  }

 private:
  std::list<Actuator> actuators_;
};

#ifndef JOSHUA_NODE_TEST
int main(int argc, char* argv[]) {
  return ros2_utils::RunNode<ActionSubscriber>(argc, argv, "actuator_subscriber");
}

#else
std::shared_ptr<rclcpp::Node> MakeActuatorSubscriberForTest(const config::Config& config) {
  return ros2_utils::CreateValidatedNode(
      "actuator_command_test",
      1,
      config,
      [](const std::string& name, int id, const config::Config& validated_config) {
        return std::make_shared<ActionSubscriber>(name, id, validated_config);
      });
}
#endif

File: ros2/actuator_subscriber.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float32

from config.proto import config_pb2
from robot.action.factory import action_factory
from robot.action.proto import action_pb2
from ros2.node_runner import run_node
from ros2.proto import ros2_data_type_pb2
from ros2.utils.packet_parser import (
    PacketParseError,
    action_packet_from_float,
    denormalize_action_packet,
    device_id_from_topic,
)
from ros2.utils.qos_setting import create_qos_setting


class ActuatorEntry:
    def __init__(self, topic, interface, limits, normalized, device_id):
        self.topic = topic
        self.interface = interface
        self.limits = limits
        self.normalized = normalized
        self.device_id = device_id
        self.subscription = None
        self.callback = None


class ActionSubscriber(Node):
    def __init__(self, node_name: str, node_id: int, config: config_pb2.Config):
        super().__init__(node_name)
        self._actuators = []

        for single_action in config.robot.actions.single_actions:
            if (
                single_action.action_type == action_pb2.ActionType.ACTUATOR
                and int(single_action.node.id) != node_id
            ):
                continue

            actuator_proto = single_action.actuator
            qos_setting = single_action.node.qos_setting
            device_id = actuator_proto.actuator_name

            try:
                interface = action_factory.create_action(single_action)
            except Exception as exc:
                self.get_logger().error(
                    f"Failed to create action interface for actuator "
                    f"'{actuator_proto.actuator_name}': {exc}"
                )
                continue

            subscription_count = 0
            for subscription in single_action.node.subscriptions:
                topic = subscription.topic
                data_type = subscription.ros2_data_type

                if data_type != ros2_data_type_pb2.FLOAT32:
                    self.get_logger().error(
                        "Unsupported ros2_data_type %s for actuator '%s'. "
                        "Only FLOAT32 is supported.",
                        str(data_type),
                        topic,
                    )
                    continue

                try:
                    topic_device_id = device_id_from_topic(topic)
                except PacketParseError as exc:
                    self.get_logger().error(
                        f"Invalid actuator Float32 topic '{topic}': {exc}"
                    )
                    continue
                if topic_device_id != device_id:
                    self.get_logger().error(
                        f"Actuator topic '{topic}' device_id '{topic_device_id}' "
                        f"does not match actuator_name '{device_id}'."
                    )
                    continue

                entry = ActuatorEntry(
                    topic=topic,
                    interface=interface,
                    limits=(
                        actuator_proto.operational_lower_limit,
                        actuator_proto.operational_upper_limit,
                    ),
                    normalized=subscription.normalized,
                    device_id=device_id,
                )
                entry.callback = self._make_float32_callback(entry)
                entry.subscription = self.create_subscription(
                    Float32,
                    entry.topic,
                    entry.callback,
                    create_qos_setting(qos_setting),
                )
                self._actuators.append(entry)
                subscription_count += 1

            if subscription_count:
                self.get_logger().info(
                    f"Configured actuator '{actuator_proto.actuator_name}' "
                    f"with {subscription_count} subscription(s)."
                )

        if not self._actuators:
            self.get_logger().error(
                f"No actuators found in configuration for node_id {node_id}!"
            )
            return

        self.get_logger().info(
            f"Actuator subscriber node started with {len(self._actuators)} "
            f"subscriptions for node_id {node_id}!"
        )

    def _make_float32_callback(self, entry: ActuatorEntry):
        """Callback for /<device_id>/<action_type> Float32 actuator topics."""

        def callback(msg: Float32):
            try:
                packet = action_packet_from_float(
                    msg.data,
                    entry.topic,
                    normalized=entry.normalized,
                )
                lower, upper = entry.limits
                packet = denormalize_action_packet(packet, lower, upper)
                entry.interface.set_action(packet)
            except PacketParseError as exc:
                self.get_logger().error(
                    f"Failed to parse Float32 payload for '{entry.topic}': {exc}"
                )
            except Exception as exc:
                self.get_logger().error(
                    f"Failed to set action for actuator '{entry.topic}': {exc}"
                )

        return callback

    def shutdown(self):
        for actuator in self._actuators:
            try:
                actuator.interface.teardown()
            except Exception as exc:
                self.get_logger().error(
                    f"Failed to teardown actuator '{actuator.topic}': {exc}"
                )


def main() -> int:
    return run_node(ActionSubscriber, "actuator_subscriber")


if __name__ == "__main__":
    raise SystemExit(main())
